atc/creds/vault/manager_test.go

```diff
@@ -57,4 +67,157 @@ var _ = Describe("VaultManager", func() {
 			Expect(manager.Validate()).ToNot(BeNil())
 		})
 	})
+
+	Describe("Config", func() {
+		var config map[string]interface{}
+		var fakeVault *httptest.Server
+
+		var configErr error
+
+		BeforeEach(func() {
+			key, err := pkix.CreateRSAKey(1024)
+			Expect(err).ToNot(HaveOccurred())
+
+			ca, err := pkix.CreateCertificateAuthority(key, "", time.Now().Add(time.Hour), "", "", "", "", "vault-ca")
+			Expect(err).ToNot(HaveOccurred())
+
+			serverKey, err := pkix.CreateRSAKey(1024)
+			Expect(err).ToNot(HaveOccurred())
+
+			serverKeyBytes, err := serverKey.ExportPrivate()
+			Expect(err).ToNot(HaveOccurred())
+
+			serverName := "vault"
+
+			serverCSR, err := pkix.CreateCertificateSigningRequest(serverKey, "", []net.IP{net.ParseIP("127.0.0.1")}, []string{serverName}, "", "", "", "", "")
+			Expect(err).ToNot(HaveOccurred())
+
+			serverCert, err := pkix.CreateCertificateHost(ca, key, serverCSR, time.Now().Add(time.Hour))
+			Expect(err).ToNot(HaveOccurred())
+
+			clientKey, err := pkix.CreateRSAKey(1024)
+			Expect(err).ToNot(HaveOccurred())
+
+			clientKeyBytes, err := clientKey.ExportPrivate()
+			Expect(err).ToNot(HaveOccurred())
+
+			clientCSR, err := pkix.CreateCertificateSigningRequest(clientKey, "", nil, nil, "", "", "", "", "concourse")
+			Expect(err).ToNot(HaveOccurred())
+
+			clientCert, err := pkix.CreateCertificateHost(ca, key, clientCSR, time.Now().Add(time.Hour))
+			Expect(err).ToNot(HaveOccurred())
+
+			serverCertBytes, err := serverCert.Export()
+			Expect(err).ToNot(HaveOccurred())
+
+			clientCertBytes, err := clientCert.Export()
+			Expect(err).ToNot(HaveOccurred())
+
+			caBytes, err := ca.Export()
+			Expect(err).ToNot(HaveOccurred())
```

Comment:
I've found myself doing this kind of dance (generating a TLS cert for vault subject to the constraints that the golang TLS clients demand) several times since I've started working on concourse full-time. I've done it with [bosh interpolate](https://bosh.io/docs/variable-types/#certificate), I've done it with the `openssl` CLI and cooking up a custom `req.conf`, and just today I learned about `certstrap` (from our [own docs](https://concourse-ci.org/vault-credential-manager.html) lol). Now that you've reinvented this wheel one more time, maybe we should just make a `concourse` subcommand that generates certs specifically for vault?

Reply:
This could make sense since we have a `concourse generate-key` command, but I'm a little wary of explicitly reinventing (+ documenting and supporting) the wheel on certificate generation because that'll likely be much higher maintenance than private keys. I'd rather people just use `certstrap` since it's explicitly designed around making cert generation easy. We only did `generate-cert` because there were incompatibilities between tooling for generating and using keys across platforms/OSes/Linux distributions/etc.

IMO the general problem of cert generation difficulty actually is made worse by having wrappers hide away the complexity. At some point you're probably going to have to understand how it all works, and then un-learn what all those wrappers did for you, and whether they made the correct choices, and whether their practices were up-to-date.